crates/solver-service/src/server.rs

```diff
@@ -507,3 +543,125 @@ async fn forward_to_discovery_service(
+
+	let token_manager = solver.token_manager();
+	let pricing_service = solver.pricing();
+
+	// Calculate total input amount in USD (sum of all inputs converted to USD)
+	let mut total_input_amount_usd = Decimal::ZERO;
+	for input in &order_data.inputs {
+		// input is [token_address, amount] - extract both
+		let token_address_u256 = input[0];
+		let amount_u256 = input[1];
+
+		// Convert token address from U256 to Address
+		// Token address is stored in the last 20 bytes of the U256
+		let mut token_bytes = [0u8; 20];
+		let token_u256_bytes = token_address_u256.to_be_bytes::<32>();
+		token_bytes.copy_from_slice(&token_u256_bytes[12..32]);
+		let token_address = solver_types::Address(token_bytes.to_vec());
+
+		// Get token info
+		let token_info = token_manager
+			.get_token_info(order_data.origin_chain_id.to::<u64>(), &token_address)
+			.map_err(|e| format!("Failed to get token info: {}", e))?;
+
+		// Convert raw amount to USD using pricing service
+		let usd_amount = convert_raw_token_to_usd(
+			&amount_u256,
+			&token_info.symbol,
+			token_info.decimals,
+			pricing_service,
+		)
+		.await?;
+
+		total_input_amount_usd += Decimal::from_str(&usd_amount)
+			.map_err(|e| format!("Failed to parse input USD amount: {}", e))?;
+	}
+
+	// Calculate total output amount in USD (sum of all outputs converted to USD)
+	let mut total_output_amount_usd = Decimal::ZERO;
+	for output in &order_data.outputs {
+		// Extract token address from the last 20 bytes of the bytes32 token field
+		let mut token_bytes = [0u8; 20];
+		token_bytes.copy_from_slice(&output.token[12..32]);
+		let token_address = solver_types::Address(token_bytes.to_vec());
+
+		// Get token info
+		let token_info = token_manager
+			.get_token_info(output.chain_id.to::<u64>(), &token_address)
+			.map_err(|e| format!("Failed to get token info: {}", e))?;
+
+		// Convert raw amount to USD using pricing service
+		let usd_amount = convert_raw_token_to_usd(
+			&output.amount,
+			&token_info.symbol,
+			token_info.decimals,
+			pricing_service,
+		)
+		.await?;
+
+		total_output_amount_usd += Decimal::from_str(&usd_amount)
+			.map_err(|e| format!("Failed to parse output USD amount: {}", e))?;
+	}
+
+	// Parse execution costs from cost estimate (already in USD)
+	let execution_cost_usd = Decimal::from_str(&cost_estimate.total)
+		.map_err(|e| format!("Failed to parse execution cost: {}", e))?;
+
+	// Calculate profit: Input (USD) - Output (USD) - Costs (USD)
+	let profit_usd = total_input_amount_usd - total_output_amount_usd - execution_cost_usd;
+
+	let hundred = Decimal::new(100_i64, 0);
+	let profit_margin_decimal = (profit_usd / total_input_amount_usd) * hundred;
+
+	tracing::debug!(
+		"Profitability calculation: input=${} (USD), output=${} (USD), cost=${} (USD), profit=${} (USD), margin={}%",
+		total_input_amount_usd,
+		total_output_amount_usd,
+		execution_cost_usd,
+		profit_usd,
+		profit_margin_decimal
+	);
+
+	Ok(profit_margin_decimal)
```

Comment:
This may make sense, but could we not just return the percentage, or do we need the decimal for some reason?